### ISSUE TITLE
Where to keep `TxConfidenceTable` if not in Context?

### ISSUE THREAD

**Issue by schildbach** [bitcoinj/bitcoinj#2531]
`TxConfidenceTable` currently is the point of truth for transaction confidences. It is meant to exist only one time per network. `Transaction.getConfidence()` is only a proxy (with apparently some issues, see #1769).

Currently, `TxConfidenceTable` resides in our `Context` so we have it available whenever we need to. However, this is in the way of making the context optional and used only for testing purposes. Unfortunately, `TxConfidenceTable` is needed in production, there is quite some logic (and usually UI) depending on it.

So the question is: where could `TxConfidenceTable` go?

* Passing in `TxConfidenceTable` references whereever we need them is likely unfeasible, and is probably the reason Context was implemented in the first place. It would affect many methods of the call chain I assume.
* If we keep the table in the Context, but make Context optional at the same time, we would need to auto-create the context if it doesn't exist yet. This would create an empty `TxConfidenceTable`, possibly unexpected and thus introduce hard to detect bugs/inconsistencies. But this might actually be ok, since `TxConfidenceTable` has weak refs anyway, and has a 1000 entry cap. So apps might already have to deal with spurious "losses of confidence". Still it doesn't feel good to me.
* We could make `TxConfidenceTable` a VM-scoped singleton. Since the table is addressed by tx hash, it could actually support multiple networks at the same time (hashes don't overlap).  Still it doesn't feel good to me.
* It could go to an object we already pass around everywhere. `Network` or `NetworkParameters` come to mind, but we want to keep `Network` immutable and get rid of `NetworkParameters` eventually. Is there some omnipresent object left I oversee?
* We could also try to undo the introduction of `TxConfidenceTable` (ages ago) such that the individual transactions keep track of confidence again. But in this case we might get rid of the confidence depth first, see #877 and #1553. Also, we might need to make sure tx instances only exist once per transaction, otherwise we'd have multiple confidences for the same tx. Our quest to make txns immutable might come in handy here?

This ticket is meant for discussing the various options.

**Comment by msgilligan** [comment]
Why not `PeerGroup`? It looks like it may have been there originally:
```
    /** Use "Context.get().getConfidenceTable()" instead */
    @Deprecated @Nullable
    public TxConfidenceTable getMemoryPool() {
        return Context.get().getConfidenceTable();
    }
```
https://github.com/bitcoinj/bitcoinj/blob/0645a835de8a5001665513d6e27e2c6517a27662/core/src/main/java/org/bitcoinj/core/PeerGroup.java#L1642-L1646

**Comment by schildbach** [comment]
> Why not `PeerGroup`? 

Good idea! However a PeerGroup doesn't always exist. One can instantiate a single Peer and receive events which should (potentially) have an effect on transaction confidences.

**Comment by msgilligan** [comment]
> Good idea! However a PeerGroup doesn't always exist.

It could be possible to share it between Peers in a PeerGroup and fallback to a private copy in a standalone Peer.

**Comment by schildbach** [comment]
Today, I investigated into the option to put TxConfidenceTable into PeerGroup. The Peer <> PeerGroup issue could be solved via listener (Peer broadcasts a "confidence event" up to PeerGroup, which manages confidences). Or maybe one of the existing event would match the problem already.

Anyway, confidence is also needed in places where PeerGroup is far away, like Transaction and Wallet, and more. For those I don't see an easy solution. Except for using Network or NetworkParameters which is available almost everywhere.

**Comment by msgilligan** [comment]
> Today, I investigated into the option to put TxConfidenceTable into PeerGroup. The Peer <> PeerGroup issue could be solved via listener (Peer broadcasts a "confidence event" up to PeerGroup, which manages confidences). Or maybe one of the existing event would match the problem already.
> 
> Anyway, confidence is also needed in places where PeerGroup is far away, like Transaction and Wallet, and more. For those I don't see an easy solution. Except for using Network or NetworkParameters which is available almost everywhere.

How about a "Blockchain" object? (perhaps `AbstractBlockChain` or maybe an interface that `AbstractBlockChain` implements)

`Wallet` doesn't explicitly "have-a" blockchain, but effectively they do (wallets are added to blockchains not vice-versa, but in order for a wallet to know about unconfirmed transactions and transaction confidences it must be attached to a blockchain.) So there should be a way for a wallet to find its attached blockchain (if any) and access transaction confidence information from there.

`Transaction` IMO shouldn't have a dependency on confidence objects. (Aside: I think we probably want to split transaction into approximately three objects: TransactionData, TransactionMessage, and WalletTransaction.)   The immutable transaction object (TransactionData?) that we are trying to move to should certainly not have direct dependencies on confidence information or the `TxConfidenceTable`.  (See https://github.com/bitcoinj/bitcoinj/pull/2547#issuecomment-1186632532)

**Comment by msgilligan** [comment]
> > How about a "Blockchain" object?
> 
> Note that "seen" is not a product of the blockchain, but of the P2P network.

I guess I consider the mempool (or a partial mempool or a partial list of pending transactions) to be a property of a blockchain (or a blockchain network.)  If we don't want that to be part of the `AbstractBlockchain` hierarchy we could define a new object for this concept, I suppose.

I also think we want to abstract from `PeerGroup` where sensible for use cases like custodial wallets, non-custodial web-wallets, etc.

> Maybe it's a mistake to try and consolidate those two confidences (P2P based, blockchain based) in one TransactionConfidence object? Maybe we should split them …

I don't think splitting should be necessary.
 
> > Transaction IMO shouldn't have a dependency on confidence objects.
> 
> Transactions currently have some logic that depends on confidence. This then has to go elsewhere.

Yes, I think we want to move that elsewhere anyway.  (And perhaps "elsewhere" is a `WalletTransaction` object)

**Comment by micheal-swiggs** [comment]
So I'm guessing that we need to remove `Transaction.getConfidence()` somehow?

**Comment by msgilligan** [comment]
> * So I'm guessing that we need to remove `Transaction.getConfidence()` somehow?

Yes. I think we should do this after creating a transaction interface and multiple (I propose 3) implementations: one for raw transaction data, one for transaction messages, and a third for wallet transactions. (There might also need to be some kind of transaction builder.) Our current transaction is used for all three roles and this (in my opinion) results in too much complexity and too many dependencies in one object.

It _might_ be OK for to have a `.getConfidence()` method on a wallet transaction object.

I think we also need to define an object or interface that represents blockchain + mempool state.  Currently this full state seems to be in a combination of existing objects including `Context`. I suppose this new object might also be limited to tracking the state of transactions for a single wallet (or maybe keychain, keybag, etc)  See also https://github.com/bitcoinj/bitcoinj/issues/1874 about creating a wallet interface and wallet module.
 
> * The logic in `Peer` for updating `TxTableConfidences` could be moved out to a listener...

I'd need to look at this a little more closely, but this could be a step we could take before deciding on or implementing some of the larger architectural changes/refactorings that I am proposing.

**Comment by micheal-swiggs** [comment]
> I think we also need to define an object or interface that represents blockchain + mempool state. Currently this full state seems to be in a combination of existing objects including `Context`. I suppose this new object might also be limited to tracking the state of transactions for a single wallet (or maybe keychain, keybag, etc) See also #1874 about creating a wallet interface and wallet module.

I somewhat agree with this. I mean specifically for confidences I'm guessing we are only interested in confidences for transactions in our wallets. For transactions already past the `EVENT_HORIZON` there is no need to track them. That leaves the transactions outside of the wallets and before the `EVENT_HORIZON`, I guess these aren't really supported today and don't need to be? Maybe no one is using BitcoinJ to track transactions outside of wallets...

**Comment by msgilligan** [comment]
> our wallets

> Maybe no one is using BitcoinJ to track transactions outside of wallets...

Remember that a **bitcoinj** `Wallet` is currently a non-custodial SPV wallet that is serialized via ProtoBuf. There are already people (including me) who are using **bitcoinj** to implement other types of wallets and non-wallet applications.

One example would be a light wallet that uses a server to track transaction confidence but uses **bitcoinj** objects in it's GUI and a **bitcoinj** HD keychain for signing transactions.

I can definitely imagine other wallets (or blockchain explorer apps) that might want to track transaction confidence.

**Comment by micheal-swiggs** [comment]
> I can definitely imagine other wallets (or blockchain explorer apps) that might want to track transaction confidence.

Completely agree with this. I guess I was a bit concerned that the new implementation might only track confidences for transactions in a `single wallet (or maybe keychain, keybag, etc)`.

**Comment by micheal-swiggs** [comment]
Just having a browse and there seems like for the time being at least that `TxConfidenceTable` could be made a Singleton on itself? It uses basic construction within `Context`... For future uses I don't think this would work since the calculation for `depth` would introduce a dependency to `AbstractBlockChain` so that we could get the current block height. The current block height could be captured by either making `TxConfidenceTable` a `NewBestBlockListener` or making `AbstractBlockChain.getChainHead().getHeight()` a `Supplier<Integer>` for `TxConfidenceTable`. My preference is for the second approach.

**Comment by msgilligan** [comment]
I haven't forgotten about this issue and am hoping we can move forward with it eventually.

* https://github.com/bitcoinj/bitcoinj/pull/3189 is a WIP (or proof-of-concept) that shows how the current implementation can really slow down (by 30x) a test/application that tries to read a complete blockchain.
* https://github.com/bitcoinj/bitcoinj/pull/3190 is a minor refactoring that may make further changes easier.

**Comment by msgilligan** [comment]
I submitted more related PRs:

* PR #3194
* PR #3195
* PR #3196
* PR #3199
* PR #3200
* PR #3201

**Comment by msgilligan** [comment]
@schildbach Did you see this comment from 2022?

https://github.com/bitcoinj/bitcoinj/issues/2531#issuecomment-1189490729

I think we need some kind of object that represents the "live" state of a P2P Bitcoin/Blockchain network. As a first step it can contain the `TxConfidenceTable`, but it could also contain other dynamic information about the state of the chain. A wallet that is connected to a network would have a reference to an instance as would a `PeerGroup`. I'm not sure what to call it, but for now I think I'll call it `PeerNetwork` (meaning the network of _all_ the peers for mainnet, or testnet, etc.) -- but I'm open to other names.

`Transaction` is trickier, but as we migrate to immutable transactions we have been considering various architectures for managing changing network state of transactions and these should ultimately use `PeerNetwork`.

**Comment by msgilligan** [comment]
> > Why not `PeerGroup`?
> 
> Good idea! However a PeerGroup doesn't always exist. One can instantiate a single Peer and receive events which should (potentially) have an effect on transaction confidences.

@schildbach:

Are there applications that manage transactions and confidences with just a single `Peer`? Do we really need to support having `TxConfidenceTable` for solo Peers?

Also, when is a `TxConfidenceTable` required? Which operations should fail when a `TxConfidenceTable` is not available?

At the very least, we know that nothing in `org.bitcoinj.base` should require a `TxConfidenceTable`. 

We can probably agree that ideally, `Transaction` _shouldn't_ need to know_ about `TxConfidenceTable` and if it weren't for `Transaction.getConfidence()` it wouldn't have to. It would also be reasonable to have `Transaction.getConfidence()` throw an `IllegalStateException` when `TxConfidenceTable` is not available _as long as_ we clearly define what an application needs to do to create a `TxConfidenceTable`.

Since `TxConfidenceTable` is only updated by `Peer` instances, it seems reasonable to say that no `TxConfidenceTable` exists until a  `PeerGroup` (or possibly a single `Peer`)  is created. 

And as far as I can tell, other than tests nothing really uses `TransactionConfidence` except classes in the `o.b.wallet` package.

**Comment by msgilligan** [comment]
I'm thinking we should agree on an architectural model with some basic constraints so that we can keep things simple:

1. There is no `TxConfidenceTable` until a `PeerGroup` is instantiated. (`IllegalStateException` if violated)
2. Wallets and applications should use PeerGroup-_associated_ (details can vary) objects to lookup `TransactionConfidence`.
3. Legacy code that can't explicitly reference a `TxConfidenceTable` automatically/transparently uses a singleton `TxConfidenceTable` based upon `Network` to load `TranactionConfidence` -- provided the table exists.
4. All use of `Context` by applications is deprecated and has no effect. `Context` becomes package-private in the `o.b..core` package.

What the implementation guarantees:

1. No or minimal code changes for typical applications (e.g. Bitcoin Wallet)
2. Behavior is well-defined for applications that follow the model.

(These constraints maybe  can be relaxed for some special cases that would then require extra code and/or code changes.) 

Proposed implementation details:

1. We create a `PeerNetwork` class that in the 0.18 release has a 1-to-1 correspondence with a `PeerGroup`.  (i.e. PeerGroup has a `getPeerNetwork()` method and `PeerNetwork` has a `getPeerGroup()` method)  In future releases there can be `PeerNetwork` instances that do not have a `PeerGroup`. This is the Peergroup-_associated_ object. It gives us a migration path to a refreshed API and allow us to in the future abstract beyond `PeerGroup`.
2. We deprecate `Transaction.getConfidence()` and start migrating classes to manage access through `PeerNetwork`.  (Maybe we don't use a `@Deprecate` tag for 0.18 and just warn of deprecation in the Javadoc.)
3. If necessary we could create `PeerNetwork` implementations that use a single `Peer` or are test fixtures, etc. But ideally we keep it simple and only have the one implementation.

Why do I think this is the simplest approach that also gives us architectural flexibility? Because the current behavior is effectively the same as if `TxConfidenceTable` did live in the `PeerGroup`. It's `Peer`s that update the TxConfidenceTable and (I believe) there are no use cases that need transaction confidence that do not also have a `PeerGroup`. (If there are some use cases that use a single peer we can add support of those when we run into them.)

Why a new object and not just use `PeerGroup`?  1. We want to support wallets and applications that do not use P2P or peers at all. 2. A new object gives us the opportunity to evolve the API in other ways (e.g. create new notification mechanisms.)

**Comment by schildbach** [comment]
> Are there applications that manage transactions and confidences with just a single Peer? Do we really need to support having TxConfidenceTable for solo Peers?

I asked myself that question recently as well. BW has a "trusted peer" mode of operation (can be multiple, but realistically it will only be one – run by the user). The general expectation is that everything works just as in normal mode.

However: BW still uses PeerGroup for managing that trusted peer, it uses `.addAddress()` to add them.

**Comment by schildbach** [comment]
> There is no TxConfidenceTable until a PeerGroup is instantiated. (IllegalStateException if violated)

I just want to add that currently confidence is persisted in the wallet protobuf, and if you restore a wallet from protobuf you probably also want to restore confidence (perhaps without a PeerGroup). Perhaps we should relax the above to

> There is no **live** TxConfidenceTable until a PeerGroup is instantiated

**Comment by schildbach** [comment]
BW has this model that a wallet *always* exists – it lives in the Android Application class. And because confidence is persisted and restored with the wallet, confidence has a similar lifecycle (even though it "lives in the threads").

PeerGroup, BlockStore and BlockChain however only live for as long as the app is connected to a network. They all live in the Android Service that's responsible for syncing the block chain.

So my prelimilnary conclusion was if we bind the ConfidenceTable to an existing object, it would perhaps be the Wallet? But of course that's a very wallet-centric view.

**Comment by msgilligan** [comment]
> So my prelimilnary conclusion was if we bind the ConfidenceTable to an existing object, it would perhaps be the Wallet? But of course that's a very wallet-centric view.

I think we want a new object (i.e. `PeerNetwork`) and it should be a property of a wallet, but not require a wallet. But it needs to have an "offline" state that can be restored (from ProtoBuf or other.)  I think `PeerNetwork` should be an interface with a `SPVPeerNetwork` implementation that we can migrate existing functionality to.  The `SPVPeerNetwork` would either have an inactive (not started or perhaps also stopped) `PeerGroup` or could have a PeerGroup added dynamically.

**Comment by msgilligan** [comment]
> ... BlockStore ... however only live for as long as the app is connected to a network. 

I assume you mean the `BlockStore` object/singleton. Its file is persistant, of course.
